util/rateLimit.go

```diff
@@ -0,0 +1,470 @@
+	ret := make(chan bool)
+	select {
+	case cm.shouldDropQueries <- shouldDropQuery{
+		c:   c,
+		ret: ret,
+	}:
+		return <-ret
+	default:
+		return false
+	}
+}
+
+// Start will kick off a goroutine to consume activity from the different activity channels,
+// as well as service queries about if a given capacityQueue should drop
+func (cm *redCongestionManager) Start(ctx context.Context, wg *sync.WaitGroup) {
+	// check if the maintainer is already running to ensure there is only one routine
+	cm.runLock.Lock()
+	defer cm.runLock.Unlock()
+	if cm.running {
+		return
+	}
+	cm.running = true
+	go cm.run(ctx, wg)
+}
+
+func (cm *redCongestionManager) run(ctx context.Context, wg *sync.WaitGroup) {
+	tick := 0
+	targetRate := float64(0)
+	consumedByClient := map[ErlClient]*[]time.Time{}
+	serves := []time.Time{}
+	exit := false
+	for {
+		// first process any new events happening
+		select {
+		// consumed events -- a client has consumed capacity from a queue
+		case e := <-cm.consumed:
+			if consumedByClient[e.c] == nil {
+				ts := []time.Time{}
+				consumedByClient[e.c] = &ts
+			}
+			*(consumedByClient[e.c]) = append(*(consumedByClient[e.c]), e.t)
+		// served events -- the capacity has been totally served
+		case e := <-cm.served:
+			serves = append(serves, e.t)
+		// "should drop" queries
+		case query := <-cm.shouldDropQueries:
+			cutoff := time.Now().Add(-1 * cm.window)
+			prune(consumedByClient[query.c], cutoff)
+			query.ret <- cm.shouldDrop(targetRate, query.c, consumedByClient[query.c])
+
+		// check for context Done, and start the thread shutdown
+		case <-ctx.Done():
+			exit = true
+		}
+		tick = (tick + 1) % cm.targetRateRefreshTicks
+		// only recalculate the service rate every N ticks, because all lists must be pruned, which can be expensive
+		// also calculate if the routine is going to exit
+		if tick == 0 || exit {
+			cutoff := time.Now().Add(-1 * cm.window)
+			prune(&serves, cutoff)
+			for c := range consumedByClient {
+				if prune(consumedByClient[c], cutoff) == 0 {
+					delete(consumedByClient, c)
+				}
+			}
+			targetRate = 0
+			// targetRate is the average service rate per client per second
+			if len(consumedByClient) > 0 {
+				serviceRate := float64(len(serves)) / float64(cm.window/time.Second)
+				targetRate = serviceRate / float64(len(consumedByClient))
+			}
+		}
+		if exit {
+			cm.setTargetRate(targetRate)
+			cm.setConsumedByClient(consumedByClient)
+			cm.setServes(serves)
+			cm.stop(wg)
+			return
+		}
+	}
+}
+
+func (cm *redCongestionManager) setTargetRate(tr float64) {
+	cm.targetRate = tr
+}
+
+func (cm *redCongestionManager) setConsumedByClient(cbc map[ErlClient]*[]time.Time) {
+	cm.consumedByClient = cbc
+}
+
+func (cm *redCongestionManager) setServes(ts []time.Time) {
+	cm.serves = ts
+}
+
+func (cm *redCongestionManager) stop(wg *sync.WaitGroup) {
+	cm.runLock.Lock()
+	defer cm.runLock.Unlock()
+	cm.running = false
+	if wg != nil {
+		wg.Done()
```

Comment:
this stop function is not a pairing one to Start().
just call set running=false and wg.Done() inside the goroutine, and implement `Stop()` that cancels the context